packages/@aws-cdk/aws-codebuild/lib/source.ts

```diff
@@ -479,10 +495,12 @@ export interface CodeCommitSourceProps extends GitSourceProps {
 class CodeCommitSource extends GitSource {
   public readonly type = CODECOMMIT_SOURCE_TYPE;
   private readonly repo: codecommit.IRepository;
+  private readonly branchOrRef?: string;
 
   constructor(props: CodeCommitSourceProps) {
     super(props);
     this.repo = props.repository;
+    this.branchOrRef = props.branchOrRef;
```

Comment:
There is actually a way to get rid of this duplication.

In `GitSource`:

```ts
abstract class GitSource extends Source {
  private readonly cloneDepth?: number;
  private readonly branchOrRef?: string;

  protected constructor(props: GitSourceProps) {
    super(props);

    this.cloneDepth = props.cloneDepth;
    this.branchOrRef = props.branchOrRef;
  }

  public bind(_scope: Construct, _project: IProject): SourceConfig {
    const superConfig = super.bind(_scope, _project);
    return {
      ...superConfig,
      sourceVersion: this.branchOrRef,
      sourceProperty: {
        ...superConfig.sourceProperty,
        gitCloneDepth: this.cloneDepth,
      },
    };
  }
}
```

Then in every place that extends `GitSource` (I'm doing `CodeCommitSource`, but the other ones will be the same):

```ts
class CodeCommitSource extends GitSource {
  public readonly type = CODECOMMIT_SOURCE_TYPE;
  private readonly repo: codecommit.IRepository;

  constructor(props: CodeCommitSourceProps) {
    super(props);
    this.repo = props.repository;
  }

  public bind(_scope: Construct, project: IProject): SourceConfig {
    // https://docs.aws.amazon.com/codebuild/latest/userguide/setting-up.html
    project.addToRolePolicy(new iam.PolicyStatement({
      actions: ['codecommit:GitPull'],
      resources: [this.repo.repositoryArn],
    }));
    const superConfig = super.bind(_scope, project);
    return {
      ...superConfig,
      sourceProperty: {
        ...superConfig.sourceProperty,
        location: this.repo.repositoryCloneUrlHttp,
      },
    };
  }
}

Reply:
I removed the `branchOrRef` property from all the subclasses and moved it onto `GitSource` and `GitThirdPartySource`, then made the subclasses explicitly assign `sourceVersion: superConfig.sourceVersion`.